Refactor primitive meshing

gibletfeets opened the pull request:
# Objective

- Primitive meshing is suboptimal
- Improve primitive meshing

## Solution

- Unify all circle generation logic to be based off of the same functionality
- Allows easier modification of suboptimal circle generation logic.

---

Among a set of a few PRs to refactor and improve primitive meshing.

## Summary of changes
- 3d geometry that uses circle generation has been refactored to utilize ``CircleIterator``. Behavior has not been changed for all shapes, except for ``capsule.rs``.
  - This has resulted in a significant improvement in speed in cases such as ``torus.rs``, and worst-case, equivalent speed but (hopefully) better readability in cases such as ``capsule.rs``.
- Various variables that specified the parameters of the geometries to be created have been changed from ``u32`` to ``usize`` to keep things more in line with ``sphere.rs``, and to allow for somewhat more convenient array indexing calculations for ``capsule.rs``. ``usize`` may not be ideal, and it may be more favorable to use ``u32`` for these values instead.
- For ``capsule.rs``, more significant breaking changes have been made:
  - The ``latitudes`` variable - which after integer division by 2 specified the quantity of lines of **longitude**, has been renamed to ``stacks``, to be consistent with how ``sphere.rs`` behaves. ``stacks`` now defines the quantity of quad rows on each hemisphere. Accordingly, the setter has also been renamed.
    - For migration, the original ``latitudes`` number should be halved. If it was an odd number, this never affected capsule geometry anyway.
  - The ``longitudes`` variable - which defined the quantity of lines of **latitude**, has been renamed to ``sectors``, to also be consistent with how ``sphere.rs`` behaves. The setter for this variable has also been renamed. 
    - No other behavior has been changed, so migration should only require renaming.

gibletfeets commented:
Images displaying that the new mesh code seems to work as intended:
![meshes1](https://github.com/bevyengine/bevy/assets/9040628/d777788b-9e16-4a50-8f5b-8295438579cd)
![meshes2](https://github.com/bevyengine/bevy/assets/9040628/bd8c4f95-da35-4df3-be93-34f72937a278)

Benchmark from unmodified logic:
![benchmark](https://github.com/bevyengine/bevy/assets/9040628/35e9961e-d73b-44ef-bb01-da4642841bca)

Benchmark from changed logic, showing minimal changes (as should be expected - the functionality was simply moved.):
![bench2](https://github.com/bevyengine/bevy/assets/9040628/29ce0a77-7f3b-4373-97e4-0945414b60cd)

rodolphito commented on crates/bevy_render/src/mesh/primitives/circle_iterator.rs:
```diff
@@ -0,0 +1,68 @@
+
+mod tests {
+    #[test]
+    fn check_count() {
```
```suggestion
    fn circle_iterator_has_correct_length() {
```

rodolphito commented on crates/bevy_render/src/mesh/primitives/circle_iterator.rs:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    #[test]
+    fn check_distances() {
```
```suggestion
    fn circle_iterator_vertices_are_equidistant() {
```

rodolphito commented on crates/bevy_render/src/mesh/primitives/circle_iterator.rs:
```diff
@@ -0,0 +1,68 @@
+    }
+}
+
+mod tests {
```
```suggestion
mod tests {
    use bevy_math::Vec2;
    use super::CircleIterator;
```

gibletfeets commented:
Preemptively apologizing for the awful git history, I am still in the process of learning git.

rparrett commented on crates/bevy_render/src/mesh/primitives/circle_iterator.rs:
```diff
@@ -0,0 +1,80 @@
+            if self.wrap {
+                self.count -= 1;
+                Some(Vec2::new(
+                    (self.theta * self.count as f32).cos(),
```
Why not `sin_cos`?

gibletfeets replied:
The main goal of this step was to duplicate the existing logic and relocate it to a unified function, and then validate that it hasn't impacted performance or broken the existing meshing. I do agree that sin_cos would make sense, though.

rparrett commented:
I sort of recall this sort of refactor being discussed in discord, and I think the main optimization being the ability to pre-compute a lookup table of sin/cos values when the iterator is constructed. That would prevent some needless trig in loops in the torus/cylinder. Have you tried that?

rparrett commented on crates/bevy_render/src/mesh/primitives/circle_iterator.rs:
```diff
@@ -0,0 +1,80 @@
+
+impl CircleIterator {
+    pub(crate) fn new(count: usize, wrap: bool) -> CircleIterator {
+        if wrap {
```
I think this `if` could potentially be removed by `count: count + wrap as usize`.

gibletfeets replied:
Correct. I received some similar feedback earlier and will revise accordingly.

rparrett commented on crates/bevy_render/src/mesh/primitives/circle_iterator.rs:
```diff
@@ -0,0 +1,80 @@
+}
+
+impl CircleIterator {
+    pub(crate) fn new(count: usize, wrap: bool) -> CircleIterator {
```
A doc comment here explaining the inputs would be appreciated.

gibletfeets replied:
Noted, will revise.

rparrett commented on crates/bevy_render/src/mesh/primitives/circle_iterator.rs:
```diff
@@ -0,0 +1,80 @@
+    }
+}
+impl Iterator for CircleIterator {
+    type Item = Vec2;
```
Just opening a discussion here:

Thoughts on using `Vec2` here? Having the resulting values accessed by `.x` and `.y` seems sort of confusing.

Maybe `(f32, f32)` or perhaps `SinCos { sin: f32, cos: f32 }`?

gibletfeets replied:
`Vec2` was chosen because the first alternative to sin/cos calls that I wanted to investigate was application of a `Mat2` for rotation, which would get generated on the initialization of the iterator - which could then be subsequently cloned for creation of identical sets of unit circle points. 

If this method does not yield favorable results, I agree that something else should be used (and if it does, it would still make sense to have it return something with labels that are more intuitive than `.x`, `.y`).

gibletfeets commented:
> I sort of recall this sort of refactor being discussed in discord, and I think the main optimization being the ability to pre-compute a lookup table of sin/cos values when the iterator is constructed. That would prevent some needless trig in loops in the torus/cylinder. Have you tried that?

I have yet to try a pre-computed lookup, but it is definitely a method that is worth investigating.

rodolphito commented:
I was the one who suggested the reduction in trigonometry, the goal with this PR was to preserve existing logic and just refactor. It was not done all in one go because this way we can benchmark the refactor and the potential optimization separately

mweatherley commented:
As I recall, the optimization that @rodolphito brought up in the first place was computing a rotation matrix for the fixed step angle and then just using that to rotate around the circle by iteratively applying it; the point of that is that you only have to do trigonometry one time and the rest is just linear algebra, so it might be a little bit faster.

On the other hand, I'm not sure it's really a priority; as we discussed, there is a concern with numerical stability when using that approach, and you run the risk, for instance, of having circular meshes with duplicated vertices not closing up properly if you aren't careful.

gibletfeets commented:
So, I ended up making a few changes:

- For portions that wanted a distribution of points/(cos, sin) values that starts and ends at 0 radians, I used `circle_iterator.cycle().take(num_points + 1)`, and I removed the `wrap` flag from the iterator.
- I changed the way the iterator works. It now produces its points via a fixed rotation matrix, and it uses f64 types internally in order to improve numerical precision for calculations.

Furthermore, it turned out I had made a mistake with creating the initial benchmark. It now actually creates the meshes, and I also made it create meshes at several resolutions. I have made sure to reproduce benchmarks for the appropriate stages of my changes.

Before moving original logic to an iterator and refactoring:
![Uploading bench_no_iter.png…]()

After initial refactor
![bench_original](https://github.com/bevyengine/bevy/assets/9040628/63a0453d-8e7b-4a2d-9d89-0dc7e2c2298b)

After changing iterator to use fixed rotation matrix.
![bench_matrices](https://github.com/bevyengine/bevy/assets/9040628/0c0e667e-93b4-42aa-b9ea-243dece7e9c7)

gibletfeets commented:
- I added additional constructors ``wrapping``, ``semicircle``, and ``quarter_circle`` to ``CircleIterator``, for performing respective functions that the various refactored primitives used.

This was fairly simple and was mostly a matter of me wanting to make it more clear what was occurring when I was manually creating these forms of iterator in the various geometry implementations.

- I reorganized the code for capsule generation quite a lot. 

I basically reimplemented it in a way that is based somewhat off of the sphere implementation, and the only preserved detail from the original is the basic optimization ideas and the math for calculating UVs. The UV math is still the same and the revision should not yield notably distinct results.

There was also some confusing variable naming - ``latitudes`` controlled the number of lines of longitude around the hemispheres, and ``longitudes`` controlled the number of lines of latitude.

More importantly, ``latitudes`` is never used for the actual geometry calculations, and instead it gets divided by 2. This means that specifying values of 15 and 14 would yield the same geometry.

So, I changed the names to be reflective of the sphere implementation, and made it so ``latitudes`` becomes ``stacks``, and ``longitudes`` becomes ``sectors``. The behavior is largely the same, however, ``stacks`` is equivalent to ``latitudes/2``, so any code that specifies the resolutions for a capsule should take this into account.

atlv24 reviewed (commented):
its lovely to see so much red in the capsule file

NthTensor commented:
I'm going to close this out, it seems to have stalled. The changes are wanted, so you're more than welcome reopen.
